### ISSUE TITLE
🐛 False positive in rule `noUndeclaredVariables`

### ISSUE THREAD

**Issue by AndreiTS** [biomejs/biome#243]
### Environment information

```block
sorry cant provide it (it's too big)
```

### What happened?

Using the hook specified in https://docs.pmnd.rs/zustand/guides/auto-generating-selectors apparently raises a false positive in the rule noUndeclaredVariables: [playground](https://biomejs.dev/playground/?indentStyle=space&quoteStyle=single&jsxQuoteStyle=single&lintRules=all&code=ZQB4AHAAbwByAHQAIAB0AHkAcABlACAAVwBpAHQAaABTAGUAbABlAGMAdABvAHIAcwA8AFMAPgAgAD0AIABTACAAZQB4AHQAZQBuAGQAcwAgAHsAIABnAGUAdABTAHQAYQB0AGUAOgAgACgAKQAgAD0APgAgAGkAbgBmAGUAcgAgAFQAIAB9AAoAIAAgACAAIAA%2FACAAUwAgACYAIAB7ACAAdQBzAGUAOgAgAHsAIABbAEsAIABpAG4AIABrAGUAeQBvAGYAIABUAF0AOgAgACgAKQAgAD0APgAgAFQAWwBLAF0AIAB9ACAAfQAKACAAIAAgACAAOgAgAG4AZQB2AGUAcgA7AAoA)

### Expected result

I think in this case there should be no warning

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by denbezrukov** [comment]
The issue is the interaction between `infer T` that creates the binding `T` and the function signature `TsFunctionType` that creates its own scope. 
https://github.com/biomejs/biome/blob/106538be9c9356eb7c3d6e40481b4ca05c315d1f/crates/biome_js_semantic/src/events.rs#L307-L314

Let start with: `type A<U> = U extends {a: infer T} ? T : never`.
 
When we don't have a `TsFunctionType`, the `T` binding belongs to `TsTypeAliasDeclaration`.
I assume that if we have `infer T` inside the `TsFunctionType`  scope and it's inside `TsConditionalType`, it should propagate to the nearest `TsTypeAliasDeclaration`/`TsConditionalType`. `TsConditionalType` doesn't create a scope now.

Do I understand correctly that `infer`  can be used only inside `extends`?

**Comment by strager** [comment]
> Do I understand correctly that `infer` can be used only inside extends?

That is the only context in which I've seen `infer`, yes.

---

Note that `infer` can get wacky as it's not lexically scoped. For example, see this [quick-lint-js test case](https://github.com/quick-lint/quick-lint-js/blob/09ae31a836ed16033fcf3b6a9bcdb03e24e70deb/test/test-parse-typescript-type.cpp#L2162):
```typescript
MyType extends (OtherType extends infer T ? infer U : InnerFalse) ? OuterTrue : OuterFalse
//                         T is usable here ^^^^^^^
//                                                 U is usable here ^^^^^^^^^
```

**Comment by denbezrukov** [comment]
@strager Thank you!
Oh, it seems that we handle this case incorrectly.
[Playground](https://biomejs.dev/playground/?indentStyle=space&quoteStyle=single&jsxQuoteStyle=single&lintRules=all&code=dAB5AHAAZQAgAEEAIAA9ACAATQB5AFQAeQBwAGUAIABlAHgAdABlAG4AZABzACAAKABPAHQAaABlAHIAVAB5AHAAZQAgAGUAeAB0AGUAbgBkAHMAIABpAG4AZgBlAHIAIABUACAAPwAgAGkAbgBmAGUAcgAgAFUAIAA6ACAASQBuAG4AZQByAEYAYQBsAHMAZQApACAAPwAgAFQAIAA6ACAATwB1AHQAZQByAEYAYQBsAHMAZQAKAA%3D%3D)

Could you please explain how quick-lint-js creates scopes for TS?

**Comment by strager** [comment]
> Could you please explain how quick-lint-js creates scopes for TS?

Assuming we are parsing `A extends B ? T : F`:

Global variable: `typescript_infer_declaration_buffer: Vec<Vec<Ident>>`

Overall type parsing algorithm (relevant bits only) (this should be obvious):
1. If we see `infer`:
   1. Skip `infer`
   2. Expect an identifier
   3. Stop
2. Parse type `A`
3. If we see `extends`:
   1. Skip `extends`
   2. Parse type `B`
   3. Expect `?`
   4. Parse type `T`
   5. Expect `:`
   6. Parse type `F`

[When we parse `B`](https://github.com/quick-lint/quick-lint-js/blob/09ae31a836ed16033fcf3b6a9bcdb03e24e70deb/src/quick-lint-js/fe/parse-type.cpp#L564-L581):
1. Push a new `Vec<Ident>` onto `typescript_infer_declaration_buffer`
2. Parse type `B`
3. Pop from `typescript_infer_declaration_buffer` into a local variable `infers`

[When we see `infer X`](https://github.com/quick-lint/quick-lint-js/blob/09ae31a836ed16033fcf3b6a9bcdb03e24e70deb/src/quick-lint-js/fe/parse-type.cpp#L245-L257) (in any type, not just inside an `extends`)
1. If `typescript_infer_declaration_buffer` is empty, report Diag_TypeScript_Infer_Outside_Conditional_Type
2. Otherwise, push `X` onto the top of `typescript_infer_declaration_buffer`

[When we parse `T`](https://github.com/quick-lint/quick-lint-js/blob/09ae31a836ed16033fcf3b6a9bcdb03e24e70deb/src/quick-lint-js/fe/parse-type.cpp#L586-L593):
1. Enter a scope
2. Declare each variable of `infers`
3. Parse type `T`
5. Exit a scope

**Comment by probablyup** [comment]
Another example of code using `infer` that triggers a false positive:

```ts
type ActionPreferenceNames<Name extends PreferenceName, Actions> = {
    [ActionName in keyof Actions]?: Actions[ActionName] extends (value: infer U) => void
      ? KeysWithValuesOfType<Required<GetPreferenceValue<Name>>, Exclude<U, undefined>>
      : never;
  };

// The U variable is undeclared biome[lint/correctness/noUndeclaredVariables]
```
